cpp/common/src/codingstandards/cpp/alertreporting/DeduplicateMacroResults.qll

```diff
@@ -0,0 +1,410 @@
+  {
+    PrimaryMacroDifferentResultElementInAllInvocations() {
+      not exists(MacroInvocation inner |
+        inner.getParentInvocation() = getAnInvocation() and
+        inner.getMacro() instanceof DifferentResultElementInAllMacroInvocations
+      )
+    }
+  }
+
+  /**
+   * Convenience predicate to know when invalid macro expansions have been reported at their macro
+   * definition.
+   */
+  private predicate reported(Macro macro) {
+    macro instanceof PrimaryMacroSameResultElementInAllInvocations or
+    macro instanceof PrimaryMacroDifferentResultElementInAllInvocations
+  }
+
+  /**
+   * A macro invocation for which the target macro does not always produce a `ResultElement`, but
+   * this specific invocation of it does.
+   *
+   * This is the "primary" / most specific macro for these result elements. It also does not match
+   * `MacroInvocation`s inside of a `MacroInvocation` of a `Macro` which always produces a
+   * `ResultElement`, indicating that the real problem lies with that other `Macro` instead of with
+   * this particular invocation.
+   */
+  class IsolatedMacroExpansionWithResultElement extends ResultMacroExpansion {
+    IsolatedMacroExpansionWithResultElement() {
+      not reported(getParentInvocation*().getMacro()) and
+      not exists(MacroInvocation inner |
+        reported(inner.getMacro()) and
+        inner.getParentInvocation*() = this
+      ) and
+      not exists(ResultMacroExpansion moreSpecific |
+        moreSpecific.getResultElement() = getResultElement() and
+        moreSpecific.getParentInvocation+() = this
+      )
+    }
+  }
+
+  /**
+   * A module for generating reports across the various cases of problematic macros, problematic
+   * macro invocations.
+   *
+   * See the doc comment for the `DeduplicateMacroResults` module for more info.
+   */
+  module Report<MacroReportConfigSig<ResultElement> ReportConfig> {
+    newtype TReportResult =
+      TReportMacroResultWithSameName(PrimaryMacroSameResultElementInAllInvocations def) or
+      TReportMacroResultWithVariedName(PrimaryMacroDifferentResultElementInAllInvocations def) or
+      TReportIsolatedMacroResult(IsolatedMacroExpansionWithResultElement def) or
+      TReportNotInMacro(ResultElement def) {
+        not exists(ResultMacroExpansion macroExpansion | macroExpansion.getResultElement() = def)
+      }
+
+    /**
+     * An instance of a `ResultElement` to be reported to a user.
+     *
+     * To show a report, use the following methods:
+     *  - `report.getPrimaryElement()`
+     *  - `report.getMessage()`
+     *  - `report.getOptionalPlaceholderLocation()`
+     *  - `report.getOptionalPlaceholderMessage()`
+     *
+     * The values returned by these methods are configured by the `MacroReportConfigSig`
+     * signature parameter.
+     */
+    class ReportResult extends TReportResult {
+      string toString() { result = getMessage() }
```

Comment:
I think we should simplify the `toString()` implementation here to be a string description of the type of report result. Otherwise `Element.toString()` depends on evaluating the entire query to produce the right message, which can create potential performance problems.

Reply:
Ah, I was observing this on `grpc/grpc`, not OpenPilot - it may not reproduce on OpenPilot.